R/sim_network_bp.R

```diff
@@ -0,0 +1,124 @@
+#' Simulate a random network branching process model with a probability of
+#' infection for each contact
+#'
+#' @description
+#' Simulate a branching process on a infinite network where the contact
+#' distribution provides a function to sample the number of contacts of each
+#' individual in the simulation. Each contact is then infected with the
+#' probability of infection. The time between each infection is determined
+#' by the contact interval function.
+#'
+#' @details
+#' The contact distribution sampled takes the network effect
+#' \eqn{q(n) \sim (n + 1)p(n + 1)} where \eqn{p(n)} is the probability
+#' density function of a distribution, e.g., Poisson or Negative binomial.
+#' That is to say, the probability of having choosing a contact at random by
+#' following up a contact chooses individuals with a probability proportional
+#' to their number of contacts. The plus one is because one of the contacts
+#' was "used" to infect the person.
+#'
+#' @inheritParams sim_linelist
+#'
+#' @return A `<data.frame>` with the contact and transmission chain data.
+#' @keywords internal
+.sim_network_bp <- function(contact_distribution,
+                            contact_interval,
+                            prob_infect,
+                            config) {
+  if (is.null(config$network) ||
+      !config$network %in% c("adjusted", "unadjusted")) {
+    stop("Network incorrectly specified, check config", call. = FALSE)
+  }
+
+  # initialise data object
+  ancestor <- vector(mode = "integer", 1e5)
+  generation <- vector(mode = "integer", 1e5)
+  infected <- vector(mode = "integer", 1e5)
+  time <- vector(mode = "double", 1e5)
+
+  # store initial individual
+  ancestor[1] <- NA_integer_
+  generation[1] <- 1L
+  # 1 is infected, 0 is non-infected contact
+  infected[1] <- 1L
+  time[1] <- 0
+
+  # initialise counters
+  next_gen_size <- 1L
+  chain_generation <- 1L
+  chain_size <- 1L
+  chain_length <- 1L
+  ancestor_idx <- 1L
+  prev_ancestors <- 1L
+
+  # run loop until no more individuals are sampled
+  while (next_gen_size > 0) {
+    if (config$network == "adjusted") {
+      # sample contact distribution (excess degree distribution)
+      q <- contact_distribution(0:1e4 + 1) * (0:1e4 + 1)
+      q <- q / sum(q)
+    } else {
+      q <- contact_distribution(0:1e4)
+    }
```

Comment:
If I'm not wrong, `q` can be calculated outside the loop as it doesn't appear to be updated within it

Reply:
Good spot! Resolved in #85.